### PR TITLE
webgpu: Refactor WebGPUProgram (part 2)

### PR DESCRIPTION
**Pull request by qjia7**
Dev

To see the logs from the Cloud Build CI, please join either our [discussion](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs) or [announcement](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs-announce) mailing list.

---
This change is [Reviewable](https://reviewable.io/reviews/tensorflow/tfjs/4604)

### CODE REVIEW

**Comment by qjia7** [comment]
@lina128 @kainino0x Please take a look, thanks.

File: tfjs-backend-webgpu/src/kernels/resize_bilinear_webgpu.ts [hunk 1]
```diff
@@ -39,10 +40,16 @@ export class ResizeBilinearProgram implements WebGPUProgram {
     this.dispatch = computeDispatch(
         this.dispatchLayout, this.outputShape, this.workGroupSize);
 
-    const adjustHeight = alignCorners && newHeight > 1;
-    const adjustWidth = alignCorners && newWidth > 1;
+    this.inputShape = inputShape;
+    this.alignCorners = alignCorners;
+    this.shaderKey = `resizeblilinear${alignCorners}`;
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/resize_bilinear_webgpu.ts:45]: nit: 
```suggestion
    this.shaderKey = `resizebilinear${alignCorners}`;
```

File: tfjs-backend-webgpu/src/kernels/select_webgpu.ts [hunk 2]
```diff
@@ -25,46 +25,52 @@ export class SelectProgram implements WebGPUProgram {
...
     this.outputShape = shape;
-    const size = util.sizeFromShape(this.outputShape);
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(
         this.dispatchLayout, this.outputShape, this.workGroupSize,
         [this.workPerThread, 1, 1]);
 
+    this.cRank = cRank;
+    this.rank = rank;
+    this.shaderKey = 'select';
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/select_webgpu.ts:44]: missing info from shader key?

File: tfjs-backend-webgpu/src/kernels/slice_webgpu.ts [hunk 3]
```diff
@@ -39,14 +40,20 @@ export class SliceProgram implements WebGPUProgram {
         this.dispatchLayout, this.outputShape, this.workGroupSize,
         [this.workPerThread, 1, 1]);
 
+    this.start = start;
+    this.destSize = destSize;
+    this.shaderKey = `slice${start}${destSize}`;
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/slice_webgpu.ts:45]: same even though these are arrays

File: tfjs-backend-webgpu/src/kernels/transpose_shared_webgpu.ts [hunk 4]
```diff
@@ -33,12 +32,15 @@ export class TransposeSharedProgram implements WebGPUProgram {
       outputShape[i] = aShape[newDim[i]];
     }
     this.outputShape = outputShape;
-    this.rank = outputShape.length;
     this.dispatchLayout = {x: [0], y: [1]};
     this.dispatch = computeDispatch(
         this.dispatchLayout, this.outputShape, this.workGroupSize, [1, 1, 1]);
 
-    this.userCode = `
+    this.shaderKey = 'transposeshared';
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/transpose_shared_webgpu.ts:39]: missing info from shader key?

File: tfjs-backend-webgpu/src/kernels/transpose_webgpu.ts [hunk 5]
```diff
@@ -25,30 +25,35 @@ export class TransposeProgram implements WebGPUProgram {
...
-    const size = util.sizeFromShape(this.outputShape);
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(
         this.dispatchLayout, this.outputShape, this.workGroupSize,
         [this.workPerThread, 1, 1]);
 
-    const switched = getSwitchedCoords(newDim);
+    this.aShape = aShape;
+    this.newDim = newDim;
+    this.shaderKey = `tranpose${newDim.join(',')}`;
```
Comment (kainino0x) [inline tfjs-backend-webgpu/src/kernels/transpose_webgpu.ts:48]: nit: `.join(',')` is redundant as it's the default stringification of an array

Also size/dtype dropped from shaderKey? maybe they're redundant?

**Review by kainino0x — APPROVED** [review]
lgtm, looks like a great cleanup. a few comments

File: tfjs-backend-webgpu/src/kernels/clip_vec4_webgpu.ts [hunk 6]
```diff
@@ -24,24 +24,30 @@ import {WebGPUProgram} from './webgpu_program';
...
     this.outputShape = outputShape;
-    const size = util.sizeFromShape(this.outputShape);
     this.dispatchLayout = flatDispatchLayout(this.outputShape);
     this.dispatch = computeDispatch(
         this.dispatchLayout, this.outputShape, this.workGroupSize,
         [this.workPerThread, 1, 1]);
-    const type = getCoordsDataType(this.outputShape.length);
+    this.minVal = minVal;
+    this.maxVal = maxVal;
+    this.shaderKey = `clipvec4${minVal}${maxVal}`;
```
[the remark this answers is not in the document]
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/clip_vec4_webgpu.ts]: Good point. Unify the shaderString like `programName_${var1}_${var2}`

File: tfjs-backend-webgpu/src/kernels/select_webgpu.ts (hunk 2 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/select_webgpu.ts:44]: The `inputs` and `output` shapes are already part of the whole shader key. You can find it in `makeShaderKey` in `webgpu_program.ts`. `cRank` and `rank` are calculated from `inputs` and `output`. So we can skip them here.

File: tfjs-backend-webgpu/src/kernels/transpose_shared_webgpu.ts (hunk 4 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/transpose_shared_webgpu.ts]: Similar here.

File: tfjs-backend-webgpu/src/kernels/transpose_webgpu.ts (hunk 5 above)
Comment (qjia7) [reply tfjs-backend-webgpu/src/kernels/transpose_webgpu.ts]: Yes. They are calculated from `outputShape` and `outputShape` is already part of the whole shader key string.

**Comment by qjia7** [comment]
@kainino0x Please take another look, thanks.

File: tfjs-backend-webgpu/src/kernels/select_webgpu.ts (hunk 2 above)
Comment (kainino0x) [reply tfjs-backend-webgpu/src/kernels/select_webgpu.ts:44]: Got it, sorry I missed this since I didn't review the previous patch. SGTM.

**Review by kainino0x — APPROVED** [review]
LGTM

**Comment by na018** [comment]
After this pull request was merged I now get the error TypeError: (intermediate value)(intermediate value).resizeBilinear is not a function
TypeError: (intermediate value)(intermediate value).resizeBilinear is not a function

could this result from these changes here? @qjia7

**Comment by qjia7** [comment]
@na018 Sorry I don't reproduce it. Can you give me a detail steps to reproduce your issue?

**Pull request merged**